core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java

```diff
@@ -40,13 +44,129 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+@SuppressFBWarnings(value = "RV_EXCEPTION_NOT_THROWN",
+    justification = "Throwing Error for it to be caught by AccumuloUncaughtExceptionHandler")
 public class ThreadPools {
 
+  public static class ExecutionError extends Error {
+
+    private static final long serialVersionUID = 1L;
+
+    public ExecutionError(String message, Throwable cause) {
+      super(message, cause);
+    }
+  }
+
   private static final Logger LOG = LoggerFactory.getLogger(ThreadPools.class);
 
   // the number of seconds before we allow a thread to terminate with non-use.
   public static final long DEFAULT_TIMEOUT_MILLISECS = 180000L;
 
+  private static final ThreadPoolExecutor SCHEDULED_FUTURE_CHECKER_POOL =
+      createFixedThreadPool(1, "Scheduled Future Checker", false);
+
+  private static final ConcurrentLinkedQueue<ScheduledFuture<?>> CRITICAL_RUNNING_TASKS =
+      new ConcurrentLinkedQueue<>();
+
+  private static final ConcurrentLinkedQueue<ScheduledFuture<?>> NON_CRITICAL_RUNNING_TASKS =
+      new ConcurrentLinkedQueue<>();
+
+  private static Runnable TASK_CHECKER = new Runnable() {
+    @Override
+    public void run() {
+      while (true) {
+        Iterator<ScheduledFuture<?>> criticalTasks = CRITICAL_RUNNING_TASKS.iterator();
+        while (criticalTasks.hasNext()) {
+          checkTaskFailed(criticalTasks.next(), true);
```

Comment:
If we pass the list to the method, then it would make the remove logic in the method a bit simpler.

```suggestion
          checkTaskFailed(criticalTasks.next(), true, CRITICAL_RUNNING_TASKS);
```

Alternatively, if we make the method return true when it should be removed, then we could do the following to have the iterator remove it.

```suggestion
          if(checkTaskFailed(criticalTasks.next(), true)){
             criticalTasks.remove();
          }
```

Reply:
Addressed in 9b53ceb